### DISCUSSION TITLE
Azure WAF defaut OWASP policies are blocking my graphql queries

### DISCUSSION THREAD

**Discussion by rsdickerson** [urql-graphql/urql#1724] [Q&A] [answered]
We have deployed our GraphQL server in Azure. With WAF in prevention mode, it seems to be blocking our graphql query requests due to excess \n characters. The details of the problem are in this file: https://github.com/rsdickerson/fileshare/blob/74a7de553db64f68bd49b2ed0f9add77034efea3/PWA%20WAF%20workaround.pdf

Do others have urql working when talking to a GraphQL server on Azure with WAF in prevention mode?

I wanted to try sending the query by first removing all the extra newline characters, but since urql is adding them (at least I think that is where they are added) I can't seem to do that.

Here's an example:

My code...

export const CALENDAR_QUERY = gql`
query Calendar {calendar {id employeeId from to periods {id areaId from to publishStatus reset timeOffOpen workOpen shiftsOpen home { id  from  to } notice} assignments {id code from to workMinutes totalMinutes areaId name label date team task home duty order format pending} areas {id name}}}
`

export function useCalendarQuery() {
    console.log("useCalendarQuery()")
        //+ areaId+") ; pause = " + !areaId)
    return useQuery({
        query: CALENDAR_QUERY,
        variables: {},
        requestPolicy: "cache-and-network",
    })
}

Here's the "query" argument posted in the request with the added \n characters.

query Calendar {
  calendar {
    id
    employeeId
    from
    to
    periods {
      id
      areaId
      from
      to
      publishStatus
      reset
      timeOffOpen
      workOpen
      shiftsOpen
      home {
        id
        from
        to
        __typename
      }
      notice
      __typename
    }
    assignments {
      id
      code
      from
      to
      workMinutes
      totalMinutes
      areaId
      name
      label
      date
      team
      task
      home
      duty
      order
      format
      pending
      __typename
    }
    areas {
      id
      name
      __typename
    }
    __typename
  }
}

We can get around this issue by adding a rule to ignore the "query" argument.

Any ideas? 

Thank you so much, Scott

**Comment by rsdickerson** [comment, marked as answer]
I found that urql is calling a print command in the GraphQL library prior to sending the request. This print command is formatting the query with newlines and spacing. Azure WAF does not like that. To get past the Azure WAF rules I had to replace all the newlines with spaces. I also removed the extra spacing, but I don't think that is really necessary. I also found that I had to move the 'id' field in a type's selection fields out of the first position. That was just weird, but without doing that Azure WAF still rejected the request.

My solution was to add this code to service-worker.js.

`
const fetchWithTrimmedBody = async ({request}) => {
    let serialized = await serialize(request)
    // Trim body of \n and spaces
    serialized.body = serialized.body.replace(/\\n\s*/g, ' ')
    let deserialized = await deserialize(serialized)
    return await fetch(deserialized)
}

registerRoute(
    ({url, request, event}) => {
        return url.pathname.endsWith('/api/gql')
    },
    fetchWithTrimmedBody,
    'POST'
)
`

Here's the serialize/deserialize functions:
[request.js.txt](https://github.com/FormidableLabs/urql/files/7178120/request.js.txt)

**Reply by kitten** [reply]
Just a quick note, we had a discussion or issue on this before and the outcome was the exact same. I don't know much about that WAF, but it seems like security theatre, so if possible it may be better to replace or adapt it 😅 Then again I haven't worked with Azure before, but it may be nice if more customers complained to them, since this is a really curious default, if it really is a default

**Reply by rsdickerson** [reply]
Thanks Phil. I will complain to Microsoft for what it's worth. For now, the above solution is working for us.

**Comment by ryan-roemer** [comment]
Hi @rsdickerson and thanks for the gist with more information!

If I'm reading your error logs correctly there appear to be some very aggressive regexes in the Azure WAF rulesets for:

- `Remote Command Execution: Unix Command Injection`
- `Remote Command Execution: Windows Command Injection`
- `Restricted SQL Character Anomaly Detection (args)`

( _Take the following with a grain of salt, as I'm just reading things here and haven't worked with Azure WAF directly_ 😉  )

Looking at https://docs.microsoft.com/en-us/azure/web-application-firewall/ag/application-gateway-customize-waf-rules-portal it appears that for some rules you can customize these rules / selectively disable rules within groups? If everything here is truly a "mandatory" rule, I think you can exclude some rules and/or modify the "rule action" to maybe not trigger a failure?

If I'm reading [this](https://docs.microsoft.com/en-us/azure/web-application-firewall/ag/application-gateway-customize-waf-rules-portal#mandatory-rules) it seems the only truly "mandatory" thing is the anomaly threshold exclusion _from_ the other rules, so disabling the other rules might solve this?

If you do go down this route and find a good configuration list, we'd love to hear about it! It really seems like vanilla GraphQL queries should be able to have a straightforward path for integration with Azure WAF!

**Reply by rsdickerson** [reply]
Thanks Ryan. We used the hammer approach initially to just have WAF completely ignore the 'query' parameter till we could find a solution. I'm sure with some experimentation we could surgically remove the WAF rules causing our GraphQL requests to be rejected. We probably won't be spending time on that though since my mandate from higher up was to make our queries work with no customizations to WAF.

I will voice our complaint to Microsoft about this.
